// @tsed/common/lib/server/components/ServerLoader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const core_1 = require("@tsed/core");
const Express = require("express");
const globby = require("globby");
const Http = require("http");
const Https = require("https");
const Path = require("path");
const ts_log_debug_1 = require("ts-log-debug");
const ServerSettingsService_1 = require("../../config/services/ServerSettingsService");
const InjectorService_1 = require("../../di/services/InjectorService");
const mvc_1 = require("../../mvc");
const HandlerBuilder_1 = require("../../mvc/class/HandlerBuilder");
const LogIncomingRequestMiddleware_1 = require("../../mvc/components/LogIncomingRequestMiddleware");
const expressApplication_1 = require("../../mvc/decorators/class/expressApplication");
const httpServer_1 = require("../decorators/httpServer");
const httpsServer_1 = require("../decorators/httpsServer");
/**
 * ServerLoader provider all method to instantiate an ExpressServer.
 *
 * It provide some features :
 *
 * * [Lifecycle hooks](docs/server-loader/lifecycle-hooks.md),
 * * [Versioning Api](docs/server-loader/versioning.md),
 * * [Authentication strategy](docs/server-loader/authentication.md).
 * * [Global errors handler](docs/server-loader/global-error-handler.md),
 * * Middleware importation,
 * * Scan directory. You can specify controllers and services directory in your project,
 *
 * ```typescript
 * // In server.ts
 * import {ServerLoader, ServerSettings} from "@tsed/common";
 * import Path = require("path");
 * @ServerSettings({
 *    rootDir: Path.resolve(__dirname),
 *    port: 8000,
 *    httpsPort: 8080,
 *    mount: {
 *      "/rest": "${rootDir}/controllers/**\/*.js"
 *    }
 * })
 * export class Server extends ServerLoader {
 *
 *     $onReady(){
 *         console.log('Server started...');
 *     }
 *
 *     $onServerInitError(err){
 *         console.error(err);
 *     }
 * }
 *
 * // In app.ts
 * import Server from "./server";
 * new Server()
 *     .start()
 *     .then(() => console.log('started'))
 *     .catch(er => console.error(er));
 *
 * ```
 *
 */
ts_log_debug_1.$log.name = "TSED";
ts_log_debug_1.$log.level = "info";
class ServerLoader {
    /**
     *
     */
    constructor() {
        this.version = "4.25.0";
        this._components = [];
        this._scannedPromises = [];
        this.callHook = (key, elseFn = new Function(), ...args) => {
            const self = this;
            if (key in this) {
                ts_log_debug_1.$log.debug(`\x1B[1mCall hook ${key}\x1B[22m`);
                return self[key](...args);
            }
            return elseFn();
        };
        this._injector = new InjectorService_1.InjectorService();
        this.createExpressApplication();
        const settings = ServerSettingsService_1.ServerSettingsService.getMetadata(this);
        if (this.$onAuth) {
            ts_log_debug_1.$log.warn("The $onAuth hooks is removed. Use OverrideMiddleware method instead of. See https://goo.gl/fufBTE.");
        }
        if (settings) {
            this.setSettings(settings);
        }
    }
    /**
     *
     * @returns {Express}
     */
    createExpressApplication() {
        const expressApp = Express();
        const originalUse = expressApp.use;
        const injector = this.injector;
        expressApp.use = function (...args) {
            args = args.map(arg => {
                if (injector.has(arg)) {
                    arg = HandlerBuilder_1.HandlerBuilder.from(arg).build(injector);
                }
                return arg;
            });
            return originalUse.call(this, ...args);
        };
        this.injector.forkProvider(expressApplication_1.ExpressApplication, expressApp);
        return this;
    }
    /**
     * Create a new HTTP server with the provided `port`.
     * @returns {ServerLoader}
     */
    createHttpServer(port) {
        const httpServer = Http.createServer(this.expressApp);
        // TODO to be removed
        /* istanbul ignore next */
        httpServer.get = () => httpServer;
        this.injector.forkProvider(httpServer_1.HttpServer, httpServer);
        this.settings.httpPort = port;
        return this;
    }
    /**
     * Create a new HTTPs server.
     *
     * `options` <IHTTPSServerOptions>:
     *
     * - `port` &lt;number&gt;: Port number,
     * - `key` &lt;string&gt; | &lt;string[]&gt; | [&lt;Buffer&gt;](https://nodejs.org/api/buffer.html#buffer_class_buffer) | &lt;Object[]&gt;: The private key of the server in PEM format. To support multiple keys using different algorithms an array can be provided either as a plain array of key strings or an array of objects in the format `{pem: key, passphrase: passphrase}`. This option is required for ciphers that make use of private keys.
     * - `passphrase` &lt;string&gt; A string containing the passphrase for the private key or pfx.
     * - `cert` &lt;string&gt; | &lt;string[]&gt; | [&lt;Buffer&gt;](https://nodejs.org/api/buffer.html#buffer_class_buffer) | [&lt;Buffer[]&gt;](https://nodejs.org/api/buffer.html#buffer_class_buffer): A string, Buffer, array of strings, or array of Buffers containing the certificate key of the server in PEM format. (Required)
     * - `ca` &lt;string&gt; | &lt;string[]&gt; | [&lt;Buffer&gt;](https://nodejs.org/api/buffer.html#buffer_class_buffer) | [&lt;Buffer[]&gt;](https://nodejs.org/api/buffer.html#buffer_class_buffer): A string, Buffer, array of strings, or array of Buffers of trusted certificates in PEM format. If this is omitted several well known "root" CAs (like VeriSign) will be used. These are used to authorize connections.
     *
     * See more info on [httpsOptions](https://nodejs.org/api/tls.html#tls_tls_createserver_options_secureconnectionlistener).
     *
     * @param options Options to create new HTTPS server.
     * @returns {ServerLoader}
     */
    createHttpsServer(options) {
        const httpsServer = Https.createServer(options, this.expressApp);
        // TODO to be removed
        /* istanbul ignore next */
        httpsServer.get = () => httpsServer;
        this.injector.forkProvider(httpsServer_1.HttpsServer, httpsServer);
        this.settings.httpsPort = options.port;
        return this;
    }
    /**
     * This method let you to add a express middleware or a Ts.ED middleware like GlobalAcceptMimes.
     *
     * ```typescript
     * @ServerSettings({
     *    rootDir,
     *    acceptMimes: ['application/json'] // optional
     * })
     * export class Server extends ServerLoader {
     *     $onMountingMiddlewares(): void|Promise<any> {
     *         const methodOverride = require('method-override');
     *
     *         this.use(GlobalAcceptMimesMiddleware)
     *             .use(methodOverride());
     *
     *         // similar to
     *         this.expressApp.use(methodOverride());
     *
     *         // but not similar to
     *         this.expressApp.use(GlobalAcceptMimesMiddleware); // in this case, this middleware will not be added correctly to express.
     *
     *         return null;
     *     }
     * }
     * ```
     * @param args
     * @returns {ServerLoader}
     */
    use(...args) {
        this.expressApp.use(...args);
        return this;
    }
    /**
     * Proxy to express set
     * @param setting
     * @param val
     * @returns {ServerLoader}
     */
    set(setting, val) {
        this.expressApp.set(setting, val);
        return this;
    }
    /**
     * Proxy to express engine
     * @param ext
     * @param fn
     * @returns {ServerLoader}
     */
    engine(ext, fn) {
        this.expressApp.engine(ext, fn);
        return this;
    }
    /**
     *
     * @returns {Promise<void>}
     */
    loadSettingsAndInjector() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            ts_log_debug_1.$log.debug("Initialize settings");
            this.settings.forEach((value, key) => {
                ts_log_debug_1.$log.info(`settings.${key} =>`, value);
            });
            ts_log_debug_1.$log.info("Build services");
            return this.injector.load();
        });
    }
    /**
     *
     */
    getSettingsService() {
        return this.settings;
    }
    /**
     * Start the express server.
     * @returns {Promise<any>|Promise}
     */
    start() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const start = new Date();
            try {
                const debug = this.settings.debug;
                /* istanbul ignore next */
                if (debug && this.settings.env !== "test") {
                    ts_log_debug_1.$log.level = "debug";
                }
                yield Promise.all(this._scannedPromises);
                yield this.callHook("$onInit");
                yield this.loadSettingsAndInjector();
                ts_log_debug_1.$log.debug("Settings and injector loaded");
                yield this.loadMiddlewares();
                yield this.startServers();
                yield this.callHook("$onReady");
                yield this.injector.emit("$onServerReady");
                ts_log_debug_1.$log.info(`Started in ${new Date().getTime() - start.getTime()} ms`);
            }
            catch (err) {
                this.callHook("$onServerInitError", undefined, err);
                return Promise.reject(err);
            }
        });
    }
    /**
     * Create a new server from settings parameters.
     * @param http
     * @param settings
     * @returns {Promise<TResult2|TResult1>}
     */
    startServer(http, settings) {
        const { address, port, https } = settings;
        ts_log_debug_1.$log.debug(`Start server on ${https ? "https" : "http"}://${settings.address}:${settings.port}`);
        const promise = new Promise((resolve, reject) => {
            http.on("listening", resolve).on("error", reject);
        }).then(() => {
            const port = http.address().port;
            ts_log_debug_1.$log.info(`HTTP Server listen on ${https ? "https" : "http"}://${settings.address}:${port}`);
            return { address: settings.address, port };
        });
        http.listen(port, address);
        return promise;
    }
    /**
     * Initiliaze all servers.
     * @returns {Bluebird<U>}
     */
    startServers() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const promises = [];
            /* istanbul ignore else */
            if (this.settings.httpPort !== false) {
                const settings = this.settings.getHttpPort();
                promises.push(this.startServer(this.httpServer, Object.assign({ https: false }, settings)).then(settings => {
                    this.settings.setHttpPort(settings);
                }));
            }
            /* istanbul ignore else */
            if (this.settings.httpsPort !== false) {
                const settings = this.settings.getHttpsPort();
                promises.push(this.startServer(this.httpsServer, Object.assign({ https: true }, settings)).then(settings => {
                    this.settings.setHttpsPort(settings);
                }));
            }
            return Promise.all(promises);
        });
    }
    /**
     * Scan and imports all files matching the pattern. See the document on the [Glob](https://www.npmjs.com/package/glob)
     * pattern for more information.
     *
     * #### Example
     *
     * ```typescript
     * import {ServerLoader} from "@tsed/common";
     * import Path = require("path");
     *
     * export class Server extends ServerLoader {
     *
     *    constructor() {
     *        super();
     *
     *        let appPath = Path.resolve(__dirname);
     *
     *        this.scan(appPath + "/controllers/**\/**.js")
     *   }
     * }
     * ```
     *
     * Theses pattern scan all files in the directories controllers, services recursively.
     *
     * !> On windows on can have an issue with the Glob pattern and the /. To solve it, build your path pattern with the module Path.
     *
     * ```typescript
     * const controllerPattern = Path.join(rootDir, 'controllers','**','*.js');
     * ```
     *
     * @param patterns
     * @param endpoint
     * @returns {ServerLoader}
     */
    scan(patterns, endpoint) {
        const promises = globby.sync(ServerLoader.cleanGlobPatterns(patterns, this.settings.exclude)).map((file) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            ts_log_debug_1.$log.debug(`Import file ${endpoint}:`, file);
            try {
                const classes = yield Promise.resolve().then(() => require(file));
                this.addComponents(classes, { endpoint });
            }
            catch (er) {
                /* istanbul ignore next */
                ts_log_debug_1.$log.error(er);
                /* istanbul ignore next */
                process.exit(-1);
            }
        }));
        this._scannedPromises = this._scannedPromises.concat(promises);
        return this;
    }
    /**
     * Add classes to the components list
     * @param classes
     * @param options
     */
    addComponents(classes, options = {}) {
        classes = Object.keys(classes)
            .map(key => classes[key])
            .filter(clazz => core_1.isClass(clazz));
        const components = Object.assign(options, {
            classes
        });
        this._components = (this._components || []).concat([components]).filter(o => !!o);
        return this;
    }
    /**
     * Add classes decorated by `@Controller()` to components container.
     *
     * ### Example
     *
     * ```typescript
     * @Controller('/ctrl')
     * class MyController{
     * }
     *
     * new ServerLoader().addControllers('/rest', [MyController])
     * ```
     * ?> If the MyController class isn't decorated, the class will be ignored.
     *
     * @param {string} endpoint
     * @param {any[]} controllers
     * @returns {ServerLoader}
     */
    addControllers(endpoint, controllers) {
        return this.addComponents(controllers, { endpoint });
    }
    /**
     * Mount all controllers files that match with `globPattern` ([Glob Pattern](https://www.npmjs.com/package/glob))
     * under the endpoint. See [Versioning Rest API](docs/server-loader/versioning.md) for more informations.
     * @param endpoint
     * @param list
     * @returns {ServerLoader}
     */
    mount(endpoint, list) {
        const patterns = [].concat(list).filter((item) => {
            if (core_1.isClass(item)) {
                this.addControllers(endpoint, [item]);
                return false;
            }
            return true;
        });
        this.scan(patterns, endpoint);
        return this;
    }
    /**
     * Initialize configuration of the express app.
     */
    loadMiddlewares() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            ts_log_debug_1.$log.debug("Mount middlewares");
            this.use(LogIncomingRequestMiddleware_1.LogIncomingRequestMiddleware);
            yield this.callHook("$onMountingMiddlewares", undefined, this.expressApp);
            yield this.injector.emit("$beforeRoutesInit");
            yield this.injector.emit("$onRoutesInit", this._components);
            delete this._components; // free memory
            yield this.injector.emit("$afterRoutesInit");
            yield this.callHook("$afterRoutesInit", undefined, this.expressApp);
            // Import the globalErrorHandler
            this.use(mvc_1.GlobalErrorHandlerMiddleware);
        });
    }
    /**
     *
     */
    setSettings(settings) {
        this.settings.set(settings);
        if (this.settings.env === "test") {
            ts_log_debug_1.$log.stop();
        }
        const bind = (property, value, map) => {
            switch (property) {
                case "mount":
                    Object.keys(this.settings.mount).forEach(key => this.mount(key, value[key]));
                    break;
                case "componentsScan":
                    this.settings.componentsScan.forEach(componentDir => this.scan(componentDir));
                    break;
                case "httpPort":
                    /* istanbul ignore else */
                    if (value !== false && this.httpServer === undefined) {
                        this.createHttpServer(value);
                    }
                    break;
                case "httpsPort":
                    /* istanbul ignore else */
                    if (value !== false && this.httpsServer === undefined) {
                        this.createHttpsServer(Object.assign(map.get("httpsOptions") || {}, { port: value }));
                    }
                    break;
            }
        };
        this.settings.forEach((value, key, map) => {
            /* istanbul ignore else */
            if (value !== undefined) {
                bind(key, value, map);
            }
        });
    }
    /**
     * Return the settings configured by the decorator [@ServerSettings](api/common/server/decorators/serversettings.md).
     *
     * @ServerSettings({
     *    rootDir: Path.resolve(__dirname),
     *    port: 8000,
     *    httpsPort: 8080,
     *    mount: {
     *      "/rest": "${rootDir}/controllers/**\/*.js"
     * }
     * })
     * export class Server extends ServerLoader {
     *     $onInit(){
     *         console.log(this.settings); // {rootDir, port, httpsPort,...}
     *     }
     * }
     * ```
     *
     * @returns {ServerSettingsService}
     */
    get settings() {
        return this.injector.settings;
    }
    /**
     * Return Express Application instance.
     * @returns {core.Express}
     */
    get expressApp() {
        return this.injector.get(expressApplication_1.ExpressApplication);
    }
    /**
     * Return the InjectorService initialized by the server.
     * @returns {InjectorService}
     */
    get injectorService() {
        return this._injector;
    }
    /**
     * Return the injectorService initialized by the server.
     * @returns {InjectorService}
     */
    get injector() {
        return this._injector;
    }
    /**
     * Return Http.Server instance.
     * @returns {Http.Server}
     */
    get httpServer() {
        return this.injectorService.get(httpServer_1.HttpServer);
    }
    /**
     * Return Https.Server instance.
     * @returns {Https.Server}
     */
    get httpsServer() {
        return this.injectorService.get(httpsServer_1.HttpsServer);
    }
    /**
     *
     * @returns {any}
     * @param files
     * @param excludes
     */
    static cleanGlobPatterns(files, excludes) {
        excludes = excludes.map((s) => "!" + s.replace(/!/gi, ""));
        return []
            .concat(files)
            .map((file) => {
            if (!require.extensions[".ts"]) {
                file = file.replace(/\.ts$/i, ".js");
            }
            return Path.resolve(file);
        })
            .concat(excludes);
    }
}
tslib_1.__decorate([
    core_1.Deprecated("Removed feature. Use ServerLoader.settings"),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", []),
    tslib_1.__metadata("design:returntype", ServerSettingsService_1.ServerSettingsService)
], ServerLoader.prototype, "getSettingsService", null);
exports.ServerLoader = ServerLoader;
