// @tsed/common/lib/mvc/class/EndpointBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@tsed/core");
const SendResponseMiddleware_1 = require("../components/SendResponseMiddleware");
const HandlerBuilder_1 = require("./HandlerBuilder");
/**
 *
 */
class EndpointBuilder {
    constructor(endpoint, router) {
        this.endpoint = endpoint;
        this.router = router;
    }
    /**
     *
     */
    onRequest(endpoint, debug) {
        return (request, response, next) => {
            /* istanbul ignore else */
            if (request.id && debug) {
                request.log.debug({
                    event: "attach.endpoint",
                    target: core_1.nameOf(endpoint.target),
                    methodClass: endpoint.methodClassName,
                    httpMethod: request.method
                });
            }
            request.setEndpoint(endpoint);
            next();
        };
    }
    /**
     *
     * @param middlewares
     */
    routeMiddlewares(middlewares) {
        this.endpoint.pathsMethods.forEach(({ path, method }) => {
            if (!!method && this.router[method]) {
                this.router[method](path, ...middlewares);
            }
            else {
                const args = [path].concat(middlewares);
                this.router.use(...args);
            }
        });
        if (!this.endpoint.pathsMethods.length) {
            this.router.use(...middlewares);
        }
    }
    /**
     *
     * @returns {any[]}
     * @param injector
     */
    build(injector) {
        const endpoint = this.endpoint;
        const debug = injector.settings.debug;
        let middlewares = []
            .concat(endpoint.beforeMiddlewares)
            .concat(endpoint.middlewares)
            .concat([endpoint])
            .concat(endpoint.afterMiddlewares)
            .concat(SendResponseMiddleware_1.SendResponseMiddleware)
            .filter((item) => !!item)
            .map((middleware) => HandlerBuilder_1.HandlerBuilder.from(middleware).build(injector));
        middlewares = [this.onRequest(endpoint, debug)].concat(middlewares);
        this.routeMiddlewares(middlewares);
        return middlewares;
    }
}
exports.EndpointBuilder = EndpointBuilder;
