// @tsed/common/lib/mvc/registries/MiddlewareRegistry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@tsed/core");
const Provider_1 = require("../../di/class/Provider");
const ProviderType_1 = require("../../di/interfaces/ProviderType");
const ProviderRegistry_1 = require("../../di/registries/ProviderRegistry");
const MiddlewareType_1 = require("../interfaces/MiddlewareType");
/**
 *
 * @type {Registry<Provider<any>, Provider>}
 */
// tslint:disable-next-line: variable-name
exports.MiddlewareRegistry = ProviderRegistry_1.GlobalProviders.createRegistry(ProviderType_1.ProviderType.MIDDLEWARE, Provider_1.Provider, {
    injectable: true,
    buildable: true
});
const middlewareRegisterFn = ProviderRegistry_1.GlobalProviders.createRegisterFn(ProviderType_1.ProviderType.MIDDLEWARE);
/**
 * Add a new middleware in the `ProviderRegistry`. This middleware will be built when `InjectorService` will be loaded.
 *
 * #### Example
 *
 * ```typescript
 * import {registerMiddleware, InjectorService} from "@tsed/common";
 *
 * export default class FooMiddleware {
 *     constructor(){}
 *     use() {
 *         return "test";
 *     }
 * }
 *
 * registerMiddleware({provide: FooMiddleware});
 * // or
 * registerMiddleware(FooMiddleware);
 *
 * const injector = new InjectorService()
 * injector.load();
 *
 * const myFooService = injector.get<FooMiddleware>(FooMiddleware);
 * fooMiddleware.use(); // test
 * ```
 *
 * @param provider Provider configuration.
 */
function registerMiddleware(provider, instance) {
    middlewareRegisterFn(provider, instance);
    ProviderRegistry_1.GlobalProviders.getRegistry(ProviderType_1.ProviderType.MIDDLEWARE)
        .get(core_1.getClassOrSymbol(provider.provide || provider))
        .store.set("middlewareType", MiddlewareType_1.MiddlewareType.MIDDLEWARE);
}
exports.registerMiddleware = registerMiddleware;
/**
 * Add a new middleware in the `ProviderRegistry`. This middleware will be built when `InjectorService` will be loaded.
 *
 * #### Example
 *
 * ```typescript
 * import {registerMiddlewareError, InjectorService} from "@tsed/common";
 *
 * export default class FooMiddleware {
 *     constructor(){}
 *     use() {
 *         return "test";
 *     }
 * }
 *
 * registerMiddlewareError({provide: MyFooService});
 * // or
 * registerMiddlewareError(MyFooService);
 *
 * const injector = new InjectorService();
 * injector.load();
 *
 * const fooMiddleware = injector.get<FooMiddleware>(FooMiddleware);
 * fooMiddleware.use(); // test
 * ```
 *
 * @param provider Provider configuration.
 */
function registerMiddlewareError(provider, instance) {
    middlewareRegisterFn(provider, instance);
    ProviderRegistry_1.GlobalProviders.getRegistry(ProviderType_1.ProviderType.MIDDLEWARE)
        .get(core_1.getClassOrSymbol(provider.provide || provider))
        .store.set("middlewareType", MiddlewareType_1.MiddlewareType.ERROR);
}
exports.registerMiddlewareError = registerMiddlewareError;
