// @tsed/swagger/lib/class/OpenApiEndpointBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@tsed/core");
const utils_1 = require("../utils");
const OpenApiModelSchemaBuilder_1 = require("./OpenApiModelSchemaBuilder");
const OpenApiParamsBuilder_1 = require("./OpenApiParamsBuilder");
/** */
class OpenApiEndpointBuilder extends OpenApiModelSchemaBuilder_1.OpenApiModelSchemaBuilder {
    constructor(endpoint, endpointUrl, pathMethod, getOperationId) {
        super(endpoint.target);
        this.endpoint = endpoint;
        this.endpointUrl = endpointUrl;
        this.pathMethod = pathMethod;
        this.getOperationId = getOperationId;
        this._paths = {};
    }
    /**
     *
     * @returns {this}
     */
    build() {
        utils_1.parseSwaggerPath(this.endpointUrl, this.pathMethod.path).forEach(({ path: endpointPath, pathParams }) => {
            const builder = new OpenApiParamsBuilder_1.OpenApiParamsBuilder(this.endpoint.target, this.endpoint.methodClassName, pathParams).build();
            const path = this._paths[endpointPath] || {};
            path[this.pathMethod.method] = this.createOperation(builder);
            Object.assign(this._definitions, builder.definitions);
            this._paths[endpointPath] = path;
        });
        return this;
    }
    /**
     *
     * @param {OpenApiResponses} builderResponses
     * @returns {OpenApiResponses}
     */
    createResponses(builderResponses) {
        const responses = this.endpoint.get("responses") || {};
        core_1.deepExtends(responses, builderResponses);
        responses[this.endpoint.get("statusCode") || "200"] = { description: "Success" };
        Object.keys(responses).forEach(code => {
            responses[code] = this.createResponse(code, responses[code]);
        });
        return responses;
    }
    /**
     *
     * @returns {Operation}
     * @param builder
     */
    createOperation(builder) {
        const operationId = this.getOperationId(`${this.endpoint.targetName}.${this.endpoint.methodClassName}`);
        const security = this.endpoint.get("security") || [];
        const produces = this.endpoint.get("produces") || [];
        const consumes = this.endpoint.get("consumes") || [];
        return core_1.deepExtends({
            operationId,
            tags: [this.getTagName()],
            parameters: builder.parameters,
            consumes,
            responses: this.createResponses(builder.responses),
            security,
            produces
        }, this.endpoint.get("operation") || {});
    }
    /**
     *
     * @returns {string}
     */
    getTagName() {
        const clazz = this.endpoint.target;
        const ctrlStore = core_1.Store.from(clazz);
        const tag = ctrlStore.get("tag");
        const name = ctrlStore.get("name");
        return name || (tag && tag.name) || this.endpoint.targetName;
    }
    /**
     *
     * @param {string | number} code
     * @param options
     * @returns {Response}
     */
    createResponse(code, options) {
        const { description, headers, examples } = core_1.deepExtends(options, this.endpoint.statusResponse(code) || {});
        const response = { description, headers, examples };
        if (this.endpoint.type === undefined) {
            return response;
        }
        response.schema = this.createSchema(this.endpoint);
        return response;
    }
    /**
     *
     * @returns {}
     */
    get paths() {
        return this._paths;
    }
}
exports.OpenApiEndpointBuilder = OpenApiEndpointBuilder;
