// @tsed/common/lib/filters/registries/FilterRegistry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Provider_1 = require("../../di/class/Provider");
const ProviderType_1 = require("../../di/interfaces/ProviderType");
const ProviderRegistry_1 = require("../../di/registries/ProviderRegistry");
/**
 *
 * @type {Registry<Provider, IProvider<any>>}
 */
// tslint:disable-next-line: variable-name
exports.FilterRegistry = ProviderRegistry_1.GlobalProviders.createRegistry(ProviderType_1.ProviderType.FILTER, Provider_1.Provider, {
    injectable: true,
    buildable: true
});
/**
 * Add a new filter in the `ProviderRegistry`. This filter will be built when `InjectorService` will be loaded.
 *
 * #### Example
 *
 * ```typescript
 * import {registerFilter, InjectorService} from "@tsed/common";
 *
 * export default class MyFooFilter {
 *     constructor(){}
 *     transform() {
 *         return "test";
 *     }
 * }
 *
 * registerFilter({provide: MyFooService});
 * // or
 * registerFilter(MyFooService);
 *
 * const injector = new InjectorService();
 *
 * const myFooService = injector.get<MyFooFilter>(MyFooFilter);
 * myFooFilter.getFoo(); // test
 * ```
 *
 * @param provider Provider configuration.
 */
exports.registerFilter = ProviderRegistry_1.GlobalProviders.createRegisterFn(ProviderType_1.ProviderType.FILTER);
/**
 *
 * @type {Map<any, any>}
 */
// tslint:disable-next-line: variable-name
exports.FilterPreHandlers = new Map();
