// ts-log-debug/src/layouts/components/JsonLayout.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @module layouts
 */
/** */
var BaseLayout_1 = require("../class/BaseLayout");
var layout_1 = require("../decorators/layout");
var inpectUtils_1 = require("../utils/inpectUtils");
/**
 *
 * @private
 */
var JsonLayout = /** @class */ (function (_super) {
    __extends(JsonLayout, _super);
    function JsonLayout() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    JsonLayout.prototype.transform = function (loggingEvent, timezoneOffset) {
        var log = {
            startTime: loggingEvent.startTime,
            categoryName: loggingEvent.categoryName,
            level: loggingEvent.level.toString(),
            data: loggingEvent.data,
            context: loggingEvent.context
        };
        log.data = log.data.map(function (data) { return inpectUtils_1.formatLogData([data]); });
        return JSON.stringify(log) + (this.config["separator"] || "");
    };
    ;
    JsonLayout = __decorate([
        layout_1.Layout({ name: "json" })
    ], JsonLayout);
    return JsonLayout;
}(BaseLayout_1.BaseLayout));
exports.JsonLayout = JsonLayout;
